### PR TITLE
Reimplement DoNotReturnNullOptionals to fix edge case

### PR DESCRIPTION
**Pull request by jtraglia**
When running this check on Web3Signer, I discovered an edge case. If there's a `return null` expression in an anonymous function, it will incorrectly result in a false positive. I've added an extra test case that demonstrates it.

To fix this, I decided to reimplement the check in a way that makes more sense to me. Instead of matching each method and checking if it contains a null return, instead find all of the null returns and check if the return value is an optional value.

Also, the edge case method in Web3Signer:
* https://github.com/ConsenSys/web3signer/blob/9bae3367671357560151601749ff73173103e02a/signing/src/main/java/tech/pegasys/web3signer/signing/KeystoreFileManager.java#L99-L140

### CODE REVIEW

File: src/main/java/tech/pegasys/tools/epchecks/DoNotReturnNullOptionals.java [hunk 1]
```diff
@@ -13,56 +13,68 @@
...
     linkType = BugPattern.LinkType.NONE)
-public class DoNotReturnNullOptionals extends BugChecker implements MethodTreeMatcher {
+public class DoNotReturnNullOptionals extends BugChecker implements ReturnTreeMatcher {
+  private static final Matcher<Tree> IS_NULL_LITERAL = Matchers.kindIs(NULL_LITERAL);
 
-  private static class ReturnNullMatcher implements Matcher<Tree> {
+  @Override
+  public Description matchReturn(final ReturnTree tree, final VisitorState state) {
+    // Filter out statements that do not return null.
+    if (tree == null || !IS_NULL_LITERAL.matches(tree.getExpression(), state)) {
```
Comment (ajsutton) [inline src/main/java/tech/pegasys/tools/epchecks/DoNotReturnNullOptionals.java:45]: hmm, so this check would only flag literal `return null` but wouldn't pick up:

```java
final String value = someBoolean ? "present" : null;
return value;
```

I'm not sure if the built-in check would handle that or not though as it does require some fairly fancy inspection logic...

**Review by ajsutton — COMMENTED** [review]
I guess the other question here is whether we'd be better off just reporting the issue to errorprone and see if they can fix it so we don't have to support our own rule.

Not against this change though - just pondering options... :)

**Comment by jtraglia** [comment]
Interesting. You're right that it would only identify null literals. I didn't know there was a built-in check in errorprone. Tomorrow, I'll check out theirs to see if it's any better. If so, we can scrap this custom check. I'll also look into handling some of that fancier logic, sounds like a fun problem.

**Comment by ajsutton** [comment]
Oh that's my bad - when you said reimplementing I took it as reimplementing a built-in version. But of course the diff clearly shows changing the existing custom rule. So yeah this is definitely better than what was there then.

**Review by ajsutton — APPROVED** [review]
LGTM. Would be interesting to see what non-literal cases we can pick up but I suspect the analysis gets very difficult pretty quickly since you're basically trying to do control flow analysis.

**Comment by ajsutton** [comment]
Will merge this one as is I think since it's an improvement and then if there's further improvement possible that can be a separate PR.

**Comment by jtraglia** [comment]
Sounds good 👍 thanks!
